protocol/contracts/amo/Ramos.sol

```diff
@@ -191,207 +264,304 @@ contract RAMOS is Ownable, Pausable {
+            bptAmountIn, minQuoteTokenAmountOut, rebalancePercentageBoundLow, rebalancePercentageBoundUp,
+            postRebalanceSlippage, 1-protocolTokenBalancerPoolIndex, treasuryPriceIndex(), quoteToken
         );
+        emit RebalanceDownExit(bptAmountIn, quoteTokenAmountOut, recipient);
 
-        lastRebalanceTimeSecs = uint64(block.timestamp);
-        emit RebalanceDown(templeAmountIn, bptIn);
+        // Collect the fees on the output quote token
+        uint256 feeAmt = quoteTokenAmountOut * rebalanceFees.rebalanceExitFeeBps / BPS_PRECISION;
+        if (feeAmt != 0) {
+            quoteToken.safeTransfer(feeCollector, feeAmt);
+        }
 
-        // deposit and stake BPT
-        bptToken.safeTransfer(address(amoStaking), bptIn);
-        amoStaking.depositAndStake(bptIn);
+        if (recipient != address(this) && recipient != address(0)) {
+            uint256 quoteTokenBalance = quoteToken.balanceOf(address(this));
+            quoteToken.safeTransfer(recipient, quoteTokenBalance);
+        }
     }
 
     /**
-     * @notice Single-side deposit stable tokens into balancer pool when TEMPLE price 
-     * is below Treasury Price Floor.
-     * @param amountIn Amount of stable tokens to deposit into balancer pool
+     * @notice Rebalance up when `protocolToken` spot price is below TPI.
+     * Single-side ADD `quoteToken` into the balancer liquidity pool to raise price.
+     * Returned BPT tokens are deposited and staked into Aura for rewards using the staking contract.
+     * Ramos rebalance fees are deducted from the amount of `quoteToken` input
+     * The remainder `quoteToken` are added into the balancer pool
+     * @dev The `quoteToken` amount must be deposited into this contract first
+     * @param quoteTokenAmountIn Amount of `quoteToken` to deposit into balancer pool
      * @param minBptOut Minimum amount of BPT tokens expected to receive
      */
-    function depositStable(
-        uint256 amountIn,
+    function rebalanceUpJoin(
+        uint256 quoteTokenAmountIn,
         uint256 minBptOut
-    ) external onlyOwner whenNotPaused {
-        _validateParams(minBptOut, amountIn, maxRebalanceAmounts.stable);
+    ) external override onlyElevatedAccess whenNotPaused enoughCooldown {
+        _validateParams(minBptOut, quoteTokenAmountIn, maxRebalanceAmounts.quoteToken);
+        lastRebalanceTimeSecs = uint64(block.timestamp);
 
-        stable.safeTransfer(address(poolHelper), amountIn);
-        // stable join
-        uint256 joinTokenIndex = templeBalancerPoolIndex == 0 ? 1 : 0;
-        uint256 bptOut = poolHelper.joinPool(
-            amountIn, minBptOut, rebalancePercentageBoundUp, rebalancePercentageBoundLow,
-            templePriceFloorNumerator, postRebalanceSlippage, joinTokenIndex, stable
-        );
+        // Collect the fees from the input quote token
+        uint256 feeAmt = quoteTokenAmountIn * rebalanceFees.rebalanceJoinFeeBps / BPS_PRECISION;
+        if (feeAmt != 0) {
```

Comment:
nit: I would add FeeCollected event

Reply:
Great idea

Reply:
Done in https://github.com/TempleDAO/temple/pull/814/files